Source/JavaScriptCore/wasm/WasmStreamingCompiler.cpp

```diff
@@ -73,45 +74,81 @@ Ref<StreamingCompiler> StreamingCompiler::create(VM& vm, CompilerMode compilerMo
 
 bool StreamingCompiler::didReceiveFunctionData(FunctionCodeIndex functionIndex, const Wasm::FunctionData&)
 {
-    if (!m_plan) {
-        m_plan = adoptRef(*new LLIntPlan(m_vm, m_info.copyRef(), m_compilerMode, Plan::dontFinalize()));
-        // Plan already failed in preparation. We do not start threaded compilation.
-        // Keep Plan failed, and "finalize" will reject promise with that failure.
-        if (!m_plan->failed()) {
-            m_remainingCompilationRequests = m_info->functions.size();
-            m_threadedCompilationStarted = true;
+    if (Options::useWasmIPInt()) {
+        if (!m_ipintPlan) {
+            m_ipintPlan = adoptRef(*new IPIntPlan(m_vm, m_info.copyRef(), m_compilerMode, Plan::dontFinalize()));
+            // Plan already failed in preparation. We do not start threaded compilation.
+            // Keep Plan failed, and "finalize" will reject promise with that failure.
+            if (!m_ipintPlan->failed()) {
+                m_remainingCompilationRequests = m_info->functions.size();
+                m_threadedCompilationStarted = true;
+            }
+        }
+        if (m_threadedCompilationStarted) {
+            Ref<Plan> plan = adoptRef(*new StreamingPlan(m_vm, m_info.copyRef(), *m_ipintPlan, functionIndex, createSharedTask<Plan::CallbackType>([compiler = Ref { *this }](Plan& plan) {
+                compiler->didCompileFunction(static_cast<StreamingPlan&>(plan));
+            })));
+            ensureWorklist().enqueue(WTFMove(plan));
+        }
+    } else {
+        if (!m_llintPlan) {
+            m_llintPlan = adoptRef(*new LLIntPlan(m_vm, m_info.copyRef(), m_compilerMode, Plan::dontFinalize()));
+            // Plan already failed in preparation. We do not start threaded compilation.
+            // Keep Plan failed, and "finalize" will reject promise with that failure.
+            if (!m_llintPlan->failed()) {
+                m_remainingCompilationRequests = m_info->functions.size();
+                m_threadedCompilationStarted = true;
+            }
+        }
+        if (m_threadedCompilationStarted) {
+            Ref<Plan> plan = adoptRef(*new StreamingPlan(m_vm, m_info.copyRef(), *m_llintPlan, functionIndex, createSharedTask<Plan::CallbackType>([compiler = Ref { *this }](Plan& plan) {
+                compiler->didCompileFunction(static_cast<StreamingPlan&>(plan));
+            })));
+            ensureWorklist().enqueue(WTFMove(plan));
```

Comment:
Let's reduce duplicated code by having a local `EntryPlan* plan` and passing that to the old code below.